File: packages/evidence/src/canonical.ts
export class CanonicalizationError extends Error {
  constructor(message: string) { super(message); this.name = 'CanonicalizationError'; }
}

export function canonicalJson(value: unknown): string {
  const seen = new Set<object>();
  return serialize(value, seen, '$');
}

function serialize(value: unknown, seen: Set<object>, path: string): string {
  if (value === null) return 'null';
  if (typeof value === 'string') { if (hasUnpairedSurrogate(value)) throw new CanonicalizationError(`${path}: malformed Unicode surrogate`); return JSON.stringify(value); }
  if (typeof value === 'boolean') return value ? 'true' : 'false';
  if (typeof value === 'number') {
    if (!Number.isFinite(value)) throw new CanonicalizationError(`${path}: non-finite numbers are forbidden`);
    if (Object.is(value, -0)) return '0';
    return JSON.stringify(value);
  }
  if (typeof value === 'bigint' || typeof value === 'symbol' || typeof value === 'function' || value === undefined) {
    throw new CanonicalizationError(`${path}: unsupported value type ${typeof value}`);
  }
  if (Array.isArray(value)) {
    if (seen.has(value)) throw new CanonicalizationError(`${path}: cyclic value`);
    seen.add(value);
    const output = `[${value.map((item, index) => serialize(item, seen, `${path}[${index}]`)).join(',')}]`;
    seen.delete(value);
    return output;
  }
  if (typeof value === 'object') {
    if (seen.has(value)) throw new CanonicalizationError(`${path}: cyclic value`);
    seen.add(value);
    const record = value as Readonly<Record<string, unknown>>;
    const keys = Object.keys(record).sort(compareUnicodeCodePoints);
    const entries = keys.map((key) => `${JSON.stringify(key)}:${serialize(record[key], seen, `${path}.${key}`)}`);
    seen.delete(value);
    return `{${entries.join(',')}}`;
  }
  throw new CanonicalizationError(`${path}: unsupported value`);
}

function hasUnpairedSurrogate(value: string): boolean {
  for (let index = 0; index < value.length; index += 1) {
    const code = value.charCodeAt(index);
    if (code >= 0xd800 && code <= 0xdbff) { const next = value.charCodeAt(index + 1); if (!(next >= 0xdc00 && next <= 0xdfff)) return true; index += 1; }
    else if (code >= 0xdc00 && code <= 0xdfff) return true;
  }
  return false;
}

function compareUnicodeCodePoints(a: string, b: string): number {
  const aa = Array.from(a);
  const bb = Array.from(b);
  const length = Math.min(aa.length, bb.length);
  for (let index = 0; index < length; index += 1) {
    const ac = aa[index]?.codePointAt(0) ?? 0;
    const bc = bb[index]?.codePointAt(0) ?? 0;
    if (ac !== bc) return ac - bc;
  }
  return aa.length - bb.length;
}

File: src/l9_assurance/evidence/canonical.py
from __future__ import annotations

import json
import math
import re
from decimal import Decimal
from typing import Any


class CanonicalizationError(ValueError):
    """Raised when a value cannot be represented by l9.canonical-json/v1."""


def canonical_json(value: Any) -> str:
    return _serialize(value, set(), "$")


def _serialize(value: Any, seen: set[int], path: str) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, str):
        _validate_unicode(value, path)
        return json.dumps(value, ensure_ascii=False, separators=(",", ":"))
    if isinstance(value, int):
        return str(value)
    if isinstance(value, float):
        return _serialize_number(value, path)
    if isinstance(value, (bytes, bytearray, memoryview)):
        raise CanonicalizationError(f"{path}: unsupported value type bytes")
    if isinstance(value, (list, tuple)):
        identity = id(value)
        if identity in seen:
            raise CanonicalizationError(f"{path}: cyclic value")
        seen.add(identity)
        try:
            return "[" + ",".join(
                _serialize(item, seen, f"{path}[{index}]") for index, item in enumerate(value)
            ) + "]"
        finally:
            seen.remove(identity)
    if isinstance(value, dict):
        identity = id(value)
        if identity in seen:
            raise CanonicalizationError(f"{path}: cyclic value")
        seen.add(identity)
        try:
            if any(not isinstance(key, str) for key in value):
                raise CanonicalizationError(f"{path}: object keys must be strings")
            entries: list[str] = []
            for key in sorted(value):
                _validate_unicode(key, f"{path}.<key>")
                encoded_key = json.dumps(key, ensure_ascii=False, separators=(",", ":"))
                entries.append(f"{encoded_key}:{_serialize(value[key], seen, f'{path}.{key}')}")
            return "{" + ",".join(entries) + "}"
        finally:
            seen.remove(identity)
    raise CanonicalizationError(f"{path}: unsupported value type {type(value).__name__}")


def _serialize_number(value: float, path: str) -> str:
    if not math.isfinite(value):
        raise CanonicalizationError(f"{path}: non-finite numbers are forbidden")
    if value == 0.0:
        return "0"
    absolute = abs(value)
    representation = repr(value).lower()
    if 1e-6 <= absolute < 1e21:
        if "e" in representation:
            fixed = format(Decimal(representation), "f")
        else:
            fixed = representation
        if "." in fixed:
            fixed = fixed.rstrip("0").rstrip(".")
        return fixed
    if "e" not in representation:
        representation = format(value, ".15e").lower()
        mantissa, exponent = representation.split("e", 1)
        mantissa = mantissa.rstrip("0").rstrip(".")
        representation = f"{mantissa}e{exponent}"
    representation = re.sub(r"e([+-])0+(\d+)$", r"e\1\2", representation)
    representation = re.sub(r"e0+(\d+)$", r"e\1", representation)
    if "e" in representation and not re.search(r"e[+-]", representation):
        exponent = int(representation.split("e", 1)[1])
        if exponent >= 0:
            representation = representation.replace("e", "e+", 1)
    return representation


def _validate_unicode(value: str, path: str) -> None:
    for index, character in enumerate(value):
        code = ord(character)
        if 0xD800 <= code <= 0xDFFF:
            raise CanonicalizationError(f"{path}[{index}]: malformed Unicode surrogate")
